File: 4Sum/Solution.cpp
class Solution {
public:
    void helper(vector<int>& num, int target, int firstIdx, int secondIdx, vector<vector<int> >& allPath) {
        int left = secondIdx + 1;
        int right = num.size() - 1;
        target = target-num[firstIdx]-num[secondIdx];
        while (left < right) {
            int sum = num[left]+num[right];
            if (sum < target) {
                left++;
            } else if (sum > target) {
                right--;
            } else {
                vector<int> path(4);
                path[0] = num[firstIdx];
                path[1] = num[secondIdx];
                path[2] = num[left];
                path[3] = num[right];
                allPath.push_back(path);
                left++;
            }
        }
    }

    vector<vector<int> > fourSum(vector<int> &num, int target) {
        int size = num.size();
        vector<vector<int> > paths;
        vector<vector<int> > res;

        sort(num.begin(), num.end());
        for(int i = 0; i < size - 3; i++) {
            for(int j = i+1; j < size - 2; j++) {
                helper(num, target, i, j, paths);
            }
        }

        for(int i = 0; i < paths.size(); i++) {
            if(find(res.begin(), res.end(), paths[i]) == res.end()) {
                res.push_back(paths[i]);
            }
        }

        return res;
    }
};

File: 4Sum/Solution.rb
def helper(nums, target, f, s, res)
    l = s + 1
    r = nums.length - 1
    while l < r do
        sum = nums[f] + nums[s] + nums[l] + nums[r]
        if sum < target then
            l += 1
        elsif sum > target then
            r -= 1
        else
            path = []
            path << nums[f]
            path << nums[s]
            path << nums[l]
            path << nums[r]
            res << path
            l += 1
            while l < r && nums[l - 1] == nums[l] do
                l += 1
            end
        end
    end
end

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
    l = nums.length
    res = []
    i = 0
    nums.sort!
    while i < l - 3 do
        j = i + 1
        while j < l - 2 do
            helper(nums, target, i, j, res)
            j += 1
            while j < l && nums[j - 1] == nums[j] do
                j += 1
            end
        end
        i += 1
        while i < l && nums[i - 1] == nums[i] do 
            i += 1
        end
    end
    return res
end
